src/Symfony/Component/Security/Http/LoginLink/LoginLinkHandler.php

```diff
@@ -0,0 +1,115 @@
+    private $secret;
+    private $options;
+    private $expiredStorage;
+
+    public function __construct(UrlGeneratorInterface $urlGenerator, UserProviderInterface $userProvider, PropertyAccessorInterface $propertyAccessor, array $signatureProperties, string $secret, array $options, ?ExpiredLoginLinkStorage $expiredStorage)
+    {
+        $this->urlGenerator = $urlGenerator;
+        $this->userProvider = $userProvider;
+        $this->propertyAccessor = $propertyAccessor;
+        $this->signatureProperties = $signatureProperties;
+        $this->secret = $secret;
+        $this->options = array_merge([
+            'route_name' => null,
+            'lifetime' => 600,
+            'max_uses' => null,
+        ], $options);
+        $this->expiredStorage = $expiredStorage;
+    }
+
+    public function createLoginLink(UserInterface $user): LoginLinkDetails
+    {
+        $expiresAt = new \DateTimeImmutable(sprintf('+%d seconds', $this->options['lifetime']));
+
+        $expires = $expiresAt->format('U');
+        $parameters = [
+            'user' => $user->getUsername(),
+            'expires' => $expires,
+            'hash' => $this->computeSignatureHash($user, $expires),
+        ];
+
+        $url = $this->urlGenerator->generate(
+            $this->options['route_name'],
+            $parameters,
+            UrlGeneratorInterface::ABSOLUTE_URL
+        );
+
+        return new LoginLinkDetails($url, $expiresAt);
+    }
+
+    public function consumeLoginLink(Request $request): UserInterface
+    {
+        $username = $request->get('user');
+
+        try {
+            $user = $this->userProvider->loadUserByUsername($username);
+        } catch (UsernameNotFoundException $exception) {
+            throw new InvalidLoginLinkException('User not found.', 0, $exception);
+        }
+
+        $hash = $request->get('hash');
+        $expires = $request->get('expires');
+        if (false === hash_equals($hash, $this->computeSignatureHash($user, $expires))) {
+            throw new InvalidLoginLinkException('User has changed since link was sent.');
+        }
+
+        if ($expires < time()) {
+            throw new ExpiredLoginLinkException('Login link has expired.');
+        }
+
+        if ($this->expiredStorage && $this->options['max_uses']) {
+            $hash = $request->get('hash');
+            if ($this->expiredStorage->countUsages($hash) >= $this->options['max_uses']) {
+                throw new ExpiredLoginLinkException(sprintf('Login link can only be used "%d" times.', $this->options['max_uses']));
+            }
+
+            $this->expiredStorage->incrementUsages($hash);
+        }
+
+        return $user;
+    }
+
+    private function computeSignatureHash(UserInterface $user, int $expires): string
+    {
+        $signatureFields = [base64_encode($user->getUsername()), $expires];
+
+        foreach ($this->signatureProperties as $property) {
+            $value = $this->propertyAccessor->getValue($user, $property);
```

Comment:
I think we should add a check if the property value is a string (or null for that matter), to avoid PHP errors.

Reply:
that's a good idea - but is there a concise way to check that? We technically could allow `is_scalar($value) || $value instanceof \Stringable` ... and maybe also we need to check for a `_toString()` method since some objects will have that still but not implement stringable

Reply:
Seems like `!is_scalar($value) && !(\is_object($value) && method_exists($value, '__toString'))` is commonly used in Symfony: https://github.com/symfony/symfony/search?q=is_scalar

Reply:
be careful that `null` is not considered a scalar